pkg/kubectl/rollback.go

```diff
@@ -283,6 +269,97 @@ func (r *DaemonSetRollbacker) Rollback(obj runtime.Object, updatedAnnotations ma
 	return rollbackSuccess, nil
 }
 
+type StatefulSetRollbacker struct {
+	c clientset.Interface
+}
+
+// toRevision is a non-negative integer, with 0 being reserved to indicate rolling back to previous configuration
+func (r *StatefulSetRollbacker) Rollback(obj runtime.Object, updatedAnnotations map[string]string, toRevision int64, dryRun bool) (string, error) {
+	if toRevision < 0 {
+		return "", revisionNotFoundErr(toRevision)
+	}
+
+	ss, ok := obj.(*apps.StatefulSet)
+	if !ok {
+		return "", fmt.Errorf("passed object is not a StatefulSet: %#v", obj)
+	}
+	versionedAppsClient := versionedAppsClientV1beta1(r.c)
+	versionedExtensionsClient := versionedExtensionsClientV1beta1(r.c)
+	versionedObj, allHistory, err := controlledHistories(versionedAppsClient, versionedExtensionsClient, ss.Namespace, ss.Name, "StatefulSet")
+	if err != nil {
+		return "", fmt.Errorf("unable to find history controlled by StatefulSet %s: %v", ss.Name, err)
+	}
+
+	versionedSS, ok := versionedObj.(*appsv1beta1.StatefulSet)
+	if !ok {
+		return "", fmt.Errorf("unable to perform type assertion for StatefulSet %s", ss.Name)
+	}
+
+	if toRevision == 0 && len(allHistory) <= 1 {
+		return "", fmt.Errorf("no last revision to roll back to")
+	}
+
+	toHistory := findHistory(toRevision, allHistory)
+	if toHistory == nil {
+		return "", revisionNotFoundErr(toRevision)
+	}
+
+	if dryRun {
+		appliedSS, err := statefulset.ApplyRevision(versionedSS, toHistory)
+		if err != nil {
+			return "", err
+		}
+		return isPodTemplateConvertible(&appliedSS.Spec.Template)
+	}
+
+	// Skip if the revision already matches current StatefulSet
+	done, err := statefulset.Match(versionedSS, toHistory)
+	if err != nil {
+		return "", err
+	}
+	if done {
+		return fmt.Sprintf("%s (current template already matches revision %d)", rollbackSkipped, toRevision), nil
+	}
+
+	// Restore revision
+	if _, err = versionedAppsClient.StatefulSets(ss.Namespace).Patch(ss.Name, types.StrategicMergePatchType, toHistory.Data.Raw); err != nil {
+		return "", fmt.Errorf("failed restoring revision %d: %v", toRevision, err)
+	}
+
+	return rollbackSuccess, nil
+}
+
+func findHistory(toRevision int64, allHistory []*appsv1beta1.ControllerRevision) *appsv1beta1.ControllerRevision {
+	// Find the history to rollback to
+	var toHistory *appsv1beta1.ControllerRevision
+	if toRevision == 0 {
+		// If toRevision == 0, find the latest revision (2nd max)
+		sort.Sort(historiesByRevision(allHistory))
+		toHistory = allHistory[len(allHistory)-2]
```

Comment:
Since this is now a function, you should return an error when `len(allHistory) <= 1`. The callers already checked it, but it's safer to check it again, so that future callers won't cause a panic (`index out of range`).

Reply:
No, you have to return an error. `findHistory` finds the history of a specific revision number. If there's only one history (let's say it's rv 5), and I asked for rv 10, you should return me an error instead of blindly return me whatever history you have.

Reply:
SGTM

Reply:
Or you can check `if allHistory[0].Revision == toRevision` when the length is 1.

Reply:
i thought about this, but the original check is shorter:
```
if toRevision == 0 && len(allHistory) <= 1 {
```